src/iface/ethernet.rs

```diff
@@ -484,9 +610,64 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
         }
         Ok(emitted_any)
     }
+
+    /// Depending on `igmp_report_state` and the therein contained
+    /// timeouts, send IGMP membership reports.
+    #[cfg(all(feature = "proto-ipv4", feature = "proto-igmp"))]
+    fn igmp_egress(&mut self, timestamp: Instant) -> Result<bool> {
+        match self.inner.igmp_report_state {
+            IgmpReportState::ToSpecificQuery(version, timeout, addr) if timestamp >= timeout => {
+                if let Some(pkt) = self.inner.igmp_report_packet(version, addr) {
+                    // Send initial membership report
+                    let tx_token = self.device.transmit().ok_or(Error::Exhausted)?;
+                    self.inner.dispatch(tx_token, timestamp, pkt)?;
+                }
+
+                self.inner.igmp_report_state = IgmpReportState::Inactive;
+                Ok(true)
+            }
+            IgmpReportState::ToGeneralQuery(version, timeout, interval, addr) if timestamp >= timeout => {
+                let next_addr = {
+                    let mut next_addrs = self.inner.ipv4_mcast_groups
+                        .iter()
+                        .map(|(group_addr, &())| group_addr)
+                        .skip_while(|group_addr| *group_addr != &addr);
```

Comment:
This is quadratic isn't it? Why not store an index into `ipv4_mcast_groups` and invalidate it when changing the map?

Reply:
If by index you mean an iterator: cannot borrow from `ipv4_mcast_groups` across multiple calls… any help is appreciated.